assembl/graphql/schema.py

```diff
@@ -1,81 +1,439 @@
+        except NoResultFound:
+            return None
 
-    @staticmethod
-    def to_global_id(type, id):
-        cls = get_named_class(type)
-        return cls.uri_generic(id)
+        discussion_id = context.matchdict['discussion_id']
+        user_id = context.authenticated_userid or Everyone
+        permissions = get_permissions(user_id, discussion_id)
+        if not result.user_can(user_id, CrudPermissions.READ, permissions):
+            raise HTTPUnauthorized()
 
-    @staticmethod
-    def get_node_from_global_id(global_id, context, info, only_type=None):
-        instance = get_named_object(global_id)
-        if not instance:
-            return None
-        # Note that instance's class may be a subclass of URI's type
-        if instance and only_type:
-            assert only_type in [cls.__name__ for cls in instance.__class__.mro()],\
-                'Received not compatible node.'
-        graphene_type = info.schema.get_type(
-            instance.__class__.__name__).graphene_type
-        if not graphene_type:
-            return None
-        return instance
+        return result
+
+# For security, always use only_fields in the Meta class to be sure we don't
+# expose every fields and relations. You need at least only_fields = ('id', )
+# to take effect.
+# Auto exposing everything will automatically convert relations
+# like AgentProfile.posts_created and create dynamically the
+# object types Post, PostConnection which will conflict with those added
+# manually.
+
+
+def resolve_langstring(langstring, locale_code):
+    """If locale_code is None, return the best lang based on user prefs,
+    otherwise respect the locale_code and return the right translation or None.
+    """
+    if langstring is None:
+        return None
+
+    if locale_code is None:
+        return langstring.best_entry_in_request().value
+
+    return {models.Locale.code_for_id(locale_id): e.value
+            for locale_id, e in langstring.entries_as_dict.items()}.get(locale_code, None)
 
 
-class AgentProfile(SQLAlchemyObjectType):
+def langstring_from_input_entries(entries):
+    """Return a LangString SA object based on GraphQL LangStringEntryInput entries.
+    """
+    if entries is not None and len(entries) > 0:
+        langstring = models.LangString.create(
+            entries[0]['value'],
+            entries[0]['locale_code'])
+        for entry in entries[1:]:
+            locale_id = models.Locale.get_id_of(entry['locale_code'])
+            langstring.add_entry(
+                models.LangStringEntry(
+                    langstring=langstring,
+                    value=entry['value'],
+                    locale_id=locale_id
+                )
+            )
 
+        return langstring
+
+    return None
+
+
+class AgentProfile(SecureObjectType, SQLAlchemyObjectType):
     class Meta:
         model = models.AgentProfile
-        interfaces = (URINode, )
+        interfaces = (Node, )
```

Comment:
Yeah, it's the graphql way to have opaque id. The client doesn't need to know what the id means. The URINode you did actually added extra complexity, and I think it was not really accurate. You can have several types in GraphQL that share the same sqlalchemy type.